File: demo/healthhub/backend/app/domain/appointment.py
"""Appointment domain model with state machine ADT.

The AppointmentStatus ADT makes illegal state transitions unrepresentable.
"""

from dataclasses import dataclass
from datetime import datetime
from typing import Literal
from uuid import UUID

from typing_extensions import assert_never


# Appointment Status ADT - makes illegal states unrepresentable
@dataclass(frozen=True)
class Requested:
    """Patient requested appointment, awaiting doctor confirmation.

    Attributes:
        requested_at: Timestamp when appointment was requested
    """

    requested_at: datetime


@dataclass(frozen=True)
class Confirmed:
    """Doctor confirmed appointment with scheduled time.

    Attributes:
        confirmed_at: Timestamp when doctor confirmed the appointment
        scheduled_time: Scheduled appointment start time
    """

    confirmed_at: datetime
    scheduled_time: datetime


@dataclass(frozen=True)
class InProgress:
    """Appointment currently happening.

    Attributes:
        started_at: Timestamp when appointment began
    """

    started_at: datetime


@dataclass(frozen=True)
class Completed:
    """Appointment finished with notes.

    Attributes:
        completed_at: Timestamp when appointment was completed
        notes: Doctor's notes from the completed appointment
    """

    completed_at: datetime
    notes: str


@dataclass(frozen=True)
class Cancelled:
    """Appointment cancelled by patient, doctor, or system.

    Attributes:
        cancelled_at: Timestamp when appointment was cancelled
        cancelled_by: Actor who cancelled the appointment
        reason: Cancellation reason text
    """

    cancelled_at: datetime
    cancelled_by: Literal["patient", "doctor", "system"]
    reason: str


type AppointmentStatus = Requested | Confirmed | InProgress | Completed | Cancelled
"""Appointment status ADT with state machine transitions.

Variants:
    Requested: Initial state, awaiting doctor confirmation
    Confirmed: Doctor confirmed with scheduled time
    InProgress: Appointment currently happening
    Completed: Appointment finished (terminal state)
    Cancelled: Appointment cancelled (terminal state)
"""


@dataclass(frozen=True)
class Appointment:
    """Appointment entity with state machine.

    Immutable domain model following Effectful patterns.
    Status field uses ADT to ensure valid state transitions.

    Attributes:
        id: Unique appointment identifier
        patient_id: Patient UUID reference
        doctor_id: Doctor UUID reference
        status: Current status ADT variant
        reason: Patient's stated reason for appointment
        created_at: Timestamp when appointment was created
        updated_at: Timestamp of last status change
    """

    id: UUID
    patient_id: UUID
    doctor_id: UUID
    status: AppointmentStatus
    reason: str
    created_at: datetime
    updated_at: datetime


# State transition validation
@dataclass(frozen=True)
class TransitionSuccess:
    """Valid state transition.

    Attributes:
        new_status: The new status that was successfully transitioned to
    """

    new_status: AppointmentStatus


@dataclass(frozen=True)
class TransitionInvalid:
    """Invalid state transition.

    Attributes:
        current_status: String name of current status
        attempted_status: String name of attempted new status
        reason: Human-readable explanation of why transition is invalid
    """

    current_status: str
    attempted_status: str
    reason: str


type TransitionResult = TransitionSuccess | TransitionInvalid
"""State transition validation result ADT.

Variants:
    TransitionSuccess: Transition is valid, contains new status
    TransitionInvalid: Transition is invalid, contains error details
"""


def validate_transition(current: AppointmentStatus, new: AppointmentStatus) -> TransitionResult:
    """Validate appointment state transition.

    Valid transitions:
    - Requested → Confirmed
    - Requested → Cancelled
    - Confirmed → InProgress
    - Confirmed → Cancelled
    - InProgress → Completed
    - InProgress → Cancelled

    Args:
        current: Current appointment status
        new: Attempted new status

    Returns:
        TransitionSuccess if valid, TransitionInvalid if not
    """
    match current:
        case Requested():
            match new:
                case Confirmed() | Cancelled():
                    return TransitionSuccess(new_status=new)
                case Requested() | InProgress() | Completed():
                    current_name = type(current).__name__
                    new_name = type(new).__name__
                    return TransitionInvalid(
                        current_status=current_name,
                        attempted_status=new_name,
                        reason=f"Cannot transition from {current_name} to {new_name}",
                    )
                case _:
                    assert_never(new)

        case Confirmed():
            match new:
                case InProgress() | Cancelled():
                    return TransitionSuccess(new_status=new)
                case Requested() | Confirmed() | Completed():
                    current_name = type(current).__name__
                    new_name = type(new).__name__
                    return TransitionInvalid(
                        current_status=current_name,
                        attempted_status=new_name,
                        reason=f"Cannot transition from {current_name} to {new_name}",
                    )
                case _:
                    assert_never(new)

        case InProgress():
            match new:
                case Completed() | Cancelled():
                    return TransitionSuccess(new_status=new)
                case Requested() | Confirmed() | InProgress():
                    current_name = type(current).__name__
                    new_name = type(new).__name__
                    return TransitionInvalid(
                        current_status=current_name,
                        attempted_status=new_name,
                        reason=f"Cannot transition from {current_name} to {new_name}",
                    )
                case _:
                    assert_never(new)

        case Completed() | Cancelled():
            current_name = type(current).__name__
            new_name = type(new).__name__
            return TransitionInvalid(
                current_status=current_name,
                attempted_status=new_name,
                reason=f"Cannot transition from {current_name} to {new_name}",
            )

        case _:
            assert_never(current)

File: demo/healthhub/src/typescript/src/models/Appointment.ts
/**
 * Appointment domain model with status state machine.
 */

// Appointment status ADT (state machine)
export type AppointmentStatus =
  | Requested
  | Confirmed
  | InProgress
  | Completed
  | Cancelled

export interface Requested {
  readonly type: 'Requested'
  readonly requestedAt: Date
}

export interface Confirmed {
  readonly type: 'Confirmed'
  readonly confirmedAt: Date
  readonly scheduledTime: Date
}

export interface InProgress {
  readonly type: 'InProgress'
  readonly startedAt: Date
}

export interface Completed {
  readonly type: 'Completed'
  readonly completedAt: Date
  readonly notes: string
}

export interface Cancelled {
  readonly type: 'Cancelled'
  readonly cancelledAt: Date
  readonly cancelledBy: 'patient' | 'doctor' | 'system'
  readonly reason: string
}

// Appointment domain model
export interface Appointment {
  readonly id: string
  readonly patientId: string
  readonly doctorId: string
  readonly status: AppointmentStatus
  readonly reason: string
  readonly createdAt: Date
  readonly updatedAt: Date
}

// API response type (matches backend - flattened status)
export interface AppointmentApiResponse {
  readonly id: string
  readonly patient_id: string
  readonly doctor_id: string
  readonly status: string // "requested" | "confirmed" | "in_progress" | "completed" | "cancelled"
  readonly reason: string
  readonly created_at: string
  readonly updated_at: string
}

// Status constructors
export const requested = (requestedAt: Date = new Date()): AppointmentStatus => ({
  type: 'Requested',
  requestedAt,
})

export const confirmed = (
  scheduledTime: Date,
  confirmedAt: Date = new Date()
): AppointmentStatus => ({
  type: 'Confirmed',
  confirmedAt,
  scheduledTime,
})

export const inProgress = (startedAt: Date = new Date()): AppointmentStatus => ({
  type: 'InProgress',
  startedAt,
})

export const completed = (
  notes: string,
  completedAt: Date = new Date()
): AppointmentStatus => ({
  type: 'Completed',
  completedAt,
  notes,
})

export const cancelled = (
  cancelledBy: 'patient' | 'doctor' | 'system',
  reason: string,
  cancelledAt: Date = new Date()
): AppointmentStatus => ({
  type: 'Cancelled',
  cancelledAt,
  cancelledBy,
  reason,
})

// Type guards
export const isRequested = (status: AppointmentStatus): status is Requested =>
  status.type === 'Requested'

export const isConfirmed = (status: AppointmentStatus): status is Confirmed =>
  status.type === 'Confirmed'

export const isInProgress = (status: AppointmentStatus): status is InProgress =>
  status.type === 'InProgress'

export const isCompleted = (status: AppointmentStatus): status is Completed =>
  status.type === 'Completed'

export const isCancelled = (status: AppointmentStatus): status is Cancelled =>
  status.type === 'Cancelled'

// Check if appointment is terminal (no further transitions)
export const isTerminal = (status: AppointmentStatus): boolean =>
  status.type === 'Completed' || status.type === 'Cancelled'

// Convert API status string to ADT
const parseStatus = (statusStr: string): AppointmentStatus => {
  switch (statusStr) {
    case 'requested':
      return requested()
    case 'confirmed':
      return confirmed(new Date())
    case 'in_progress':
      return inProgress()
    case 'completed':
      return completed('')
    case 'cancelled':
      return cancelled('system', 'Unknown reason')
    default:
      return requested() // Default fallback
  }
}

// Converter from API response to domain model
export const fromApiResponse = (response: AppointmentApiResponse): Appointment => ({
  id: response.id,
  patientId: response.patient_id,
  doctorId: response.doctor_id,
  status: parseStatus(response.status),
  reason: response.reason,
  createdAt: new Date(response.created_at),
  updatedAt: new Date(response.updated_at),
})

// Display helpers
export const getStatusLabel = (status: AppointmentStatus): string => {
  switch (status.type) {
    case 'Requested':
      return 'Requested'
    case 'Confirmed':
      return 'Confirmed'
    case 'InProgress':
      return 'In Progress'
    case 'Completed':
      return 'Completed'
    case 'Cancelled':
      return 'Cancelled'
  }
}

export const getStatusColor = (status: AppointmentStatus): string => {
  switch (status.type) {
    case 'Requested':
      return '#f59e0b' // Amber
    case 'Confirmed':
      return '#3b82f6' // Blue
    case 'InProgress':
      return '#8b5cf6' // Purple
    case 'Completed':
      return '#10b981' // Green
    case 'Cancelled':
      return '#ef4444' // Red
  }
}
